### PR TITLE
Document the JSON schema of `direct_url.json`

### PR DESCRIPTION
**Pull request by edgarrmondragon**
Adds a JSON schema for the structure of `direct_url.json`.

Based on https://github.com/edgarrmondragon/pep610/blob/41a39253521637e85f364f60d1106c25dfa03d02/tests/fixtures/direct_url.schema.json with some changes to make it more accurate:

* Updated the JSON schema draft version to [2019-09](https://json-schema.org/draft/2019-09/release-notes) so `deprecated` can be used for `archive_info.hash`.
* Fixed the pattern of `archive_info.hash` to actually accomodate valid algorithm names. e.g. `md5=...` wouldn't have been valid before.

See the [discuss thread](https://discuss.python.org/t/clarify-that-the-direct-url-json-url-field-must-be-a-spec-compliant-url/46518/7) for more context.

Relevant docs preview: https://python-packaging-user-guide--1512.org.readthedocs.build/en/1512/specifications/direct-url-data-structure/#json-schema

----
📚 Documentation preview 📚: https://python-packaging-user-guide--1512.org.readthedocs.build/en/1512/

### CODE REVIEW

**Comment by brettcannon** [comment]
@edgarrmondragon FYI CI failed.

File: source/specifications/direct-url-data-structure.rst [hunk 1]
```diff
@@ -213,6 +213,128 @@ vcs command
    this field is the Subversion revision number in the corresponding
    repository.
 
+JSON Schema
+===========
+
+The following JSON Schema can be used to validate the contents of ``direct_url.json``:
+
+.. code-block::
```
Comment (brettcannon) [inline source/specifications/direct-url-data-structure.rst:221]: Can you specify the code type as JSON?

**Comment by edgarrmondragon** [comment]
> @edgarrmondragon FYI CI failed.

I don't think those failures are related to the changes here, since the scheduled runs failed too: https://github.com/pypa/packaging.python.org/actions/runs/8076025103

btw thanks for reviewing @brettcannon!

File: source/specifications/direct-url-data-structure.rst (hunk 1 above)
Comment (edgarrmondragon) [reply source/specifications/direct-url-data-structure.rst:221]: Is it necessary even with the `highlight` directive at the top of the file?

https://github.com/pypa/packaging.python.org/blob/4c86169ad719358f63632dbc93629190c35721b1/source/specifications/direct-url-data-structure.rst#L1

Comment (brettcannon) [reply source/specifications/direct-url-data-structure.rst:221]: I didn't notice that top-level notification! Sorry about that!

**Comment by brettcannon** [comment]
> I don't think those failures are related to the changes here, since the scheduled runs failed too: https://github.com/pypa/packaging.python.org/actions/runs/8076025103

They probably aren't, but if you're at all interested in helping out w/ the failures, I think the dead links are:

```
(guides/single-sourcing-package-version: line   12) broken    https://github.com/pypa/pip/blob/main/setup.py - 404 Client Error: Not Found for url: https://github.com/pypa/pip/blob/main/setup.py
(tutorials/installing-packages: line  444) broken    https://stackoverflow.com/a/14638025 - 403 Client Error: Forbidden for url: https://stackoverflow.com/a/14638025
```

**Comment by chrysle** [comment]
The failing StackOverflow link is being addressed in #1510, but judging by latest activity there, I guess a separate PR to remove the link to the now discarded `setup.py` would be welcome!

File: source/specifications/direct-url-data-structure.rst (hunk 1 above)
Comment (edgarrmondragon) [reply source/specifications/direct-url-data-structure.rst:221]: No worries!

**Review by webknjaz — APPROVED** [review]

File: source/specifications/direct-url-data-structure.rst (hunk 1 above)
Comment (webknjaz) [reply source/specifications/direct-url-data-structure.rst:221]: I verified that it's rendered as highlighted @ https://python-packaging-user-guide--1512.org.readthedocs.build/en/1512/specifications/direct-url-data-structure/#json-schema. Don't forget to mark this thread as resolved if there's no extra concerns anymore.

**Pull request merged**